src/gui/notifications/dbusnotifier.cpp

```diff
@@ -0,0 +1,94 @@
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include "dbusnotifier.h"
+
+#include <QDBusConnection>
+#include <QDBusPendingCallWatcher>
+#include <QDBusPendingReply>
+
+#include "base/global.h"
+#include "dbusnotificationsinterface.h"
+
+DBusNotifier::DBusNotifier(QObject *parent)
+    : QObject(parent)
+    , m_notificationsInterface {new DBusNotificationsInterface(u"org.freedesktop.Notifications"_qs
+                  , u"/org/freedesktop/Notifications"_qs, QDBusConnection::sessionBus())}
+{
+    // Testing for 'DBusNotificationsInterface::isValid()' isn't helpful here.
+    // If the notification daemon is configured to run 'as needed'
+    // the above check can be false if the daemon wasn't started
+    // by another application. In this case DBus will be able to
+    // start the notification daemon and complete our request. Such
+    // a daemon is xfce4-notifyd, DBus autostarts it and after
+    // some inactivity shuts it down. Other DEs, like GNOME, choose
+    // to start their daemons at the session startup and have it sit
+    // idling for the whole session.
+
+    connect(m_notificationsInterface, &DBusNotificationsInterface::ActionInvoked, this, &DBusNotifier::onActionInvoked);
+    connect(m_notificationsInterface, &DBusNotificationsInterface::NotificationClosed, this, &DBusNotifier::onNotificationClosed);
+}
+
+void DBusNotifier::showMessage(const QString &title, const QString &message, const int timeout)
+{
+    // Assign "default" action to notification to make it clickable
+    const QStringList actions {u"default"_qs, {}};
+    const QVariantMap hints {{u"desktop-entry"_qs, u"org.qbittorrent.qBittorrent"_qs}};
+    QDBusPendingReply<uint> reply = m_notificationsInterface->notify(u"qBittorrent"_qs, 0
+            , u"qbittorrent"_qs, title, message, actions, hints, timeout);
+    auto *watcher = new QDBusPendingCallWatcher(reply, this);
+    connect(watcher, &QDBusPendingCallWatcher::finished, this, [this](QDBusPendingCallWatcher *self)
+    {
+        const QDBusPendingReply<uint> reply = *self;
+        if (!reply.isError())
+        {
+            const uint messageID = reply.value();
+            m_activeMessages.insert(messageID);
+        }
+
+        self->deleteLater();
+    });
+}
+
+void DBusNotifier::onActionInvoked(const uint messageID, const QString &action)
+{
+    Q_UNUSED(action);
+
+    // Check whether the notification is sent by qBittorrent
+    // to avoid reacting to unrelated notifictions
+    if (m_activeMessages.contains(messageID))
+        emit messageClicked();
+}
+
+void DBusNotifier::onNotificationClosed(const uint messageID, const uint reason)
+{
+    Q_UNUSED(reason);
+
+    m_activeMessages.remove(messageID);
```

Comment:
So the notification server will send every close event to qbt?
Also can you point out where `DBusNotificationsInterface::NotificationClosed` is emitted?

Reply:
> So the notification server will send every close event to qbt?

Yes.


> Also can you point out where `DBusNotificationsInterface::NotificationClosed` is emitted?

Somewhere inside `QDBusAbstractInterface` implementation. When some signal of derived class is connected it tries to link it with corresponding signal of D-Bus interface.